stats/stats.go

```diff
@@ -29,45 +29,67 @@ var mtsMap = map[string]map[string]int64{}
 var dimMap = map[string]metrics.DimMap{}
 
 // ReportStats starts reporting stats on the configured time interval.
-func ReportStats(config *Config, log *logrus.Entry) {
+func ReportStats(config *Config, log *logrus.Entry) chan bool {
+	shutdown := make(chan bool, 1)
 	if config == nil || config.Interval == 0 {
-		log.Debug("Skipping stats reporting because it is configured off")
-		return
-	}
-
-	if config.Subject != "" {
-		metrics.NewCounter("", nil)
+		if log != nil {
+			log.Debug("Skipping stats reporting because it is configured off")
+		}
+		return shutdown
 	}
 
 	go func() {
-		log.WithFields(logrus.Fields{
-			"interval":      config.Interval,
-			"metric_prefix": config.Prefix,
-		}).Infof("Starting to report stats every %d seconds", config.Interval)
+		if log != nil {
+			log.WithFields(logrus.Fields{
+				"interval":      config.Interval,
+				"metric_prefix": config.Prefix,
+			}).Infof("Starting to report stats every %d seconds", config.Interval)
+		}
+
 		ticks := time.Tick(time.Duration(config.Interval) * time.Second)
-		for range ticks {
-			go func() {
-				statLock.Lock()
-				for k, series := range mtsMap {
-					for sha, val := range series {
-						dims := dimMap[sha]
-						name := config.Prefix
-						if name != "" {
-							name += "."
-						}
-						name += k
-						metrics.NewCounter(name, dims).CountN(val, nil)
-						go func(n string, v int64, d metrics.DimMap) {
-							d["value"] = v
-							log.WithFields(logrus.Fields(d)).Infof("%s = %d", n, v)
-						}(name, val, dims)
-					}
+		for {
+			select {
+			case <-shutdown:
+				if log != nil {
+					log.Info("Shutting down")
 				}
-
-				statLock.Unlock()
-			}()
+			case <-ticks:
+				go report(log, config.Prefix)
+			}
 		}
 	}()
+
+	return shutdown
+}
+
+func report(log *logrus.Entry, prefix string) {
+	results := make(map[string][]map[string]interface{})
+
+	statLock.Lock()
+	for k, series := range mtsMap {
+		for sha, val := range series {
+			dims := dimMap[sha]
+			name := prefix
+			if name != "" && !strings.HasSuffix(name, ".") {
+				name += "."
+			}
+			name += k
+			metrics.CountN(name, val, dims)
+
+			dims["value"] = val
```

Comment:
The `dimsMap` is keyed by a SHA256 of the dims without the metric name though. If you have two metrics, with the same dims, they will have the same SHA, and resolve to the same entry in `dimsMap`.

A simpler example would be *no dims*. If you had two metrics with no dims and different values, then the log message would use the same `value`.

Would it be easier if I added a failing test?